internal/locate/region_cache.go

```diff
@@ -1743,6 +1742,17 @@ func (c *RegionCache) OnRegionEpochNotMatch(bo *retry.Backoffer, ctx *RPCContext
 	newRegions := make([]*Region, 0, len(currentRegions))
 	// If the region epoch is not ahead of TiKV's, replace region meta in region cache.
 	for _, meta := range currentRegions {
+		valid, start, end, err := c.codec.DecodeRegionRange(meta.GetStartKey(), meta.GetEndKey())
+		if err != nil {
+			logutil.BgLogger().Error("failed to decode region range", zap.Error(err))
+			return false, err
+		}
+		// If cachedRegion is not valid, skip inserting it to cache.
+		if !valid {
+			continue
+		}
+		meta.StartKey = start
```

Comment:
I don't think it's safe to write the meta in place since the meta slice could be shared across the goroutine.